Sources/SymDiff/source/BoogieStructuralComparisonManager.cs

```diff
@@ -0,0 +1,269 @@
+            blockA.TransferCmd.GetType() != blockB.TransferCmd.GetType())
+            return false;
+
+        if (blockA.TransferCmd is GotoCmd t1 &&
+            blockB.TransferCmd is GotoCmd t2)
+        {
+            foreach (var (blkTarget1, blkTarget2) in t1.labelTargets.Zip(t2.labelTargets))
+            {
+                if (!blockMapping.ContainsKey(blkTarget1))
+                    blockMapping.Add(blkTarget1, blkTarget2);
+                else if (blockMapping[blkTarget1] != blkTarget2)
+                    return false;
+            }
+        }
+
+        var i = 0;
+        var j = 0;
+        while (i < blockA.Cmds.Count && j < blockB.Cmds.Count)
+        {
+            if (IsIrrelevantCommand(blockA.Cmds[i]))
+            {
+                i++;
+                continue;
+            }
+            if (IsIrrelevantCommand(blockB.Cmds[j]))
+            {
+                j++;
+                continue;
+            }
+            if (!Compare(blockA.Cmds[i], blockB.Cmds[j]))
+                return false;
+            i++;
+            j++;
+        }
+        // Any remaining commands must be irrelevant
+        for (; i < blockA.Cmds.Count; ++i)
+        {
+            if (!IsIrrelevantCommand(blockA.Cmds[i]))
+                return false;
+        }
+        for (; j < blockB.Cmds.Count; ++j)
+        {
+            if (!IsIrrelevantCommand(blockB.Cmds[j]))
+                return false;
+        }
+
+        return true;
+    }
+
+    private bool EqualsStructuralExpr(Expr exprA, Expr exprB)
+    {
+        var combinedMapping = localVarMapping.Concat(GlobalVarMapping).ToDictionary();
+        var exprComparator = new ExprComparatorWithRenaming(combinedMapping);
+        var res = exprComparator.Compare(exprA, exprB);
+        var updatedMapping = exprComparator.GetUpdatedVariableMapping;
+
+        // update the mappings
+        foreach (var (v1, v2) in updatedMapping)
+        {
+            switch (v1)
+            {
+                case GlobalVariable:
+                case Constant:
+                    GlobalVarMapping.TryAdd(v1, v2);
+                    break;
+                default:
+                    localVarMapping.TryAdd(v1, v2);
+                    break;
+            };
+        }
+
+        return res;
+    }
+    
+    private bool Compare(Cmd? cmdA, Cmd? cmdB)
+    {
+        if (cmdA == null || cmdB == null) return cmdA == cmdB;
+        if (cmdA.GetType() != cmdB.GetType()) return false;
+        return (cmdA, cmdB) switch
+        {
+            (AssertCmd a, AssertCmd b) => EqualsStructuralExpr(a.Expr, b.Expr),
+            (AssumeCmd a, AssumeCmd b) => EqualsStructuralExpr(a.Expr, b.Expr),
+            (AssignCmd a, AssignCmd b) => EqualsStructuralAssign(a, b),
+            (CallCmd   a, CallCmd   b) => EqualsStructuralCall(a, b),
+            (HavocCmd  a, HavocCmd  b) => EqualsStructuralHavoc(a, b),
+            _ => throw new Exception(
```

Comment:
This could also be a `NotSupportedException`. I'm halfway inclined to say that this should just be a warning followed by `return false`. The idea there is that we should be able to analyze arbitrary code, no matter how complex or what language features it uses, and at least be able to sometimes compare it for syntactic equality. Anything unsupported isn't really fatal. It just needs to be handled by the semantic check. And, even there, reporting "unknown" for unsupported features will make the whole pipeline more robust.